src/main.zig

```diff
@@ -119,12 +122,79 @@ fn rocRepl(allocator: Allocator, opt: RocOpt, args: []const []const u8) !void {
     fatal("not implemented", .{});
 }
 
-fn rocFormat(allocator: Allocator, opt: RocOpt, args: []const []const u8) !void {
-    _ = allocator;
+fn rocFormat(allocator: Allocator, args: []const []const u8) !void {
+    const roc_file_path = if (args.len > 0) args[0] else "main.roc";
 
-    std.debug.print("TODO roc format\n{}\n{s}\n\n", .{ opt, args });
+    const input_file = try std.fs.cwd().openFile(roc_file_path, .{ .mode = .read_only });
+    defer input_file.close();
 
-    fatal("not implemented", .{});
+    const contents = try input_file.reader().readAllAlloc(allocator, std.math.maxInt(usize));
+    defer allocator.free(contents);
+
+    var module_env = base.ModuleEnv.init(allocator);
+    defer module_env.deinit();
+
+    var parse_ast = parse.parse(&module_env, allocator, contents);
+    defer parse_ast.deinit();
+
+    var formatter = fmt.init(parse_ast, allocator);
+    defer formatter.deinit();
+
+    const formatted_output = formatter.formatFile();
+    defer allocator.free(formatted_output);
+
+    const output_file = try std.fs.cwd().createFile(roc_file_path, .{});
+    defer output_file.close();
+    try output_file.writeAll(formatted_output);
+}
+
+test "format single file" {
+    const allocator = std.testing.allocator;
+    const filename = "test.roc";
+    {
+        const roc_file = try std.fs.cwd().createFile(filename, .{});
+        defer roc_file.close();
+        try roc_file.writeAll(
+            \\module []
+            \\
+            \\foo =      "bar"
+        );
+    }
+    defer std.fs.cwd().deleteFile(filename) catch std.debug.panic("Failed to clean up temp file", .{});
+
+    // TODO: Share build step between tests
+    {
+        const build_result = try std.process.Child.run(.{
+            .allocator = allocator,
+            .argv = &[_][]const u8{ "zig", "build" },
+        });
+        defer allocator.free(build_result.stdout);
+        defer allocator.free(build_result.stderr);
+        try std.testing.expectEqual(build_result.term.Exited, 0);
+    }
+
+    {
+        const format_result = try std.process.Child.run(.{
+            .allocator = allocator,
+            .argv = &[_][]const u8{ "./zig-out/bin/roc", "format", filename },
+        });
```

Comment:
I think you can just call roc_format with the arena and `&.{"format", filename}` instead of building the binary